src/wp-includes/media.php

```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+		$img_url = $uploads_dir['baseurl'] . '/' . $img_file;
+	}
+
+	$img_sizes = $img_meta['sizes'];
+
+	// Add full size to the img_sizes array.
+	$img_sizes['full'] = array(
+		'width'  => $img_meta['width'],
+		'height' => $img_meta['height'],
+		'file'   => wp_basename( $img_meta['file'] )
+	);
+
+	// Calculate the image aspect ratio.
+	$img_ratio = $img_height / $img_width;
+
+	/*
+	 * Images that have been edited in WordPress after being uploaded will
+	 * contain a unique hash. Look for that hash and use it later to filter
+	 * out images that are leftovers from previous versions.
+	 */
+	$img_edited = preg_match( '/-e[0-9]{13}/', $img_url, $img_edit_hash );
+
+	// Set up an array to hold the matched image sources.
+	$sources = array();
+
+	/*
+	 * Loop through available images and only use images that are resized
+	 * versions of the same rendition.
+	 */
+	foreach ( $img_sizes as $img ) {
+
+		// Filter out images that are leftovers from previous renditions.
+		if ( $img_edited && ! strpos( $img['file'], $img_edit_hash[0] ) ) {
+			continue;
+		}
+
+		// Calculate the new image ratio.
+		$img_ratio_compare = $img['height'] / $img['width'];
+
+		// If the new ratio differs by less than 0.01, use it.
+		if ( abs( $img_ratio - $img_ratio_compare ) < 0.01 ) {
+			$sources[ $img['width'] ] = path_join( dirname( $img_url ), $img['file'] ) . ' ' . $img['width'] .'w';
+		}
+	}
+
+	/**
+	 * Filter the output of wp_get_srcset_array().
+	 *
+	 * @since 4.4.0
+	 *
+	 * @param array        $sources  An array of image urls and widths.
+	 * @param int          $id       Attachment ID for image.
+	 * @param array|string $size     Size of image, either array or string.
+	 */
+	return apply_filters( 'wp_attachment_img_srcset', $sources, $id, $size );
+}
+
+/**
+ * Get the value for the 'srcset' attribute.
+ *
+ * @since 4.4.0
+ *
+ * @param int    $id   Image attachment ID.
+ * @param string $size Optional. Name of image size. Default value: 'medium'.
+ * @return string|bool A 'srcset' value string or false.
+ */
+function wp_attachment_img_srcset( $id, $size = 'medium' ) {
+	$srcset_array = wp_attachment_img_srcset_array( $id, $size );
+
+	// Only return a srcset value if there is more than one source.
+	if ( count( $srcset_array ) <= 1 ) {
+		return false;
+	}
+
+	return implode( ', ', $srcset_array );
+}
+
+/**
+ * Return a source size attribute for an image from an array of values.
+ *
+ * @since 4.4.0
+ *
+ * @param int    $id   Image attachment ID.
+ * @param string $size Optional. Name of image size. Default value: 'medium'.
+ * @param array  $args {
+ *     Optional. Arguments to retrieve posts.
+ *
+ *     @type array|string $sizes An array or string containing of size information.
+ *     @type int          $width A single width value used in the default `sizes` string.
+ * }
+ * @return string|bool A valid source size value for use in a 'sizes' attribute or false.
+ */
+function wp_attachment_img_sizes( $id, $size = 'medium', $args = null ) {
```

Comment:
If $args is more defined than "extra arguments," we can name it appropriately.  Is it $sizes?